src/Symfony/Component/DependencyInjection/ContainerBuilder.php

```diff
@@ -407,6 +408,27 @@ public function fileExists(string $path, $trackContents = true): bool
         return $exists;
     }
 
+    final public function willBeAvailable(string $package, string $class, string $method = null): bool
```

Comment:
the $method arg seems unused. IMHO `willBeAvailable('x/y', y::class) && method_exists(y::class, 'new')` is more clear, by letting the user own the extra condition.

Reply:
even the $class check is debatable, if we'd check the package for a certain version (`willBeAvailable('x/y', '1.0')`), as it implies the actual state

perhaps at this point we prefer a crash, rather than a silent skip

Reply:
checking a class is better than checking a version:

- checking version is not compatible with replaced packages (such packages are returning true for `InstalledVersions::isInstalled($package)` but don't return a version for `InstalledVersions::getVersion()`)
- feature detection is more flexible than detecting package version (the frontend ecosystem learned a lot from that, by moving to feature detection rather than sniffing the browser version and assuming the related support). Also, it works even for dev versions

Reply:
i generally agree. I also missed `!class_exists(InstalledVersions::class)` which explains why we need both a package check and a feature check.

Compared to the frontend ecosystem. im not sure checking N browsers for 1 feature vs. 1 package for 1 feature compares well, or justifies not checking for packages in our case.

Reply:
Due to `replace` rules in composer, you could have any package actually be responsible for shipping the class. And so you cannot get the version number introducing it.
We do have one *main* package owning the class (which is why we can have a package name here), and composer is handling replacement for us in this case. But not for versions.

Reply:
`replace` rules take a version constraint though :/

ive no experience with eg. `"replace": { "x/y": ">=2"}`, only `* / self.version`, however conceptually id say we know the replaced versions for a package

no intention to push this direction :) current solution is OK.

from a maintainer pov i pick `is x/y 2.0 minimally available` any day compared to awkward feature checking hoping it pinpoints to the right version.

Reply:
Good idea, I removed the `$method` argument.

Versions checks are no-go on my side: feature checks are much more maintainable than cryptic version numbers. Thanks for proposing.